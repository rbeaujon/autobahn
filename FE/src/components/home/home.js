import React, { useEffect, useState } from "react";
import {GetSessionApi} from "../../services/API/getSession.api";
import {registerData} from "../../services/API/registerData.api";
import {SectorsAPI} from "../../services/API/sectors.api";
import Regex from "../../utils/helpers/regex";
import { v4 as uuidv4 } from 'uuid';

import "./home.styles.scss";



export const Home = () => {

const[sectors, setSectors] = useState([]);
const[verification, setVerification] = useState({name: false, sector: false, terms: false, error: false, chars: false})
const [session, setSession] = useState(JSON.parse(sessionStorage.getItem('session')) || null);
const[isRegistered, setIsRegistered] = useState(false);
const[isDisabled, setIsDisabled] = useState(true);

const getSectors = async () => {
  const data = await SectorsAPI();
  setSectors(data);
}

const sendToRegister = async (data) => {
  const header = {
    method: 'POST',
    headers: {
    'Accept': 'application/json',
    'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  };
  const response = await registerData(header)
  if(response.status === 201){
    console.log(`Register successfully`);
    document.getElementById("terms").checked = false;
    setIsRegistered(true)
  }else{
    console.error('Unavailable error: ' + response.status);
  }
  handleVerification();
}

//Get the Sectors available on DB
useEffect(() => {
  getSectors();
  
  if(session){
    const header = {
      method: 'POST',
      headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
      },
      body: JSON.stringify({session})
    };

    const getSession = async () => {
     
      const response = await GetSessionApi(header); //fieldId,session,userId
      if(response.status === 200 && response.data.length !== 0){
        const sector = document.getElementById("sector");
        document.getElementById("name").value = response.data[0].userId;
        response.data[0].fieldId.split(",").map(num => parseInt(num)).map(elem => sector[elem].selected = true)
        document.getElementById("terms").checked = false
        handleVerification()
      }  
    }
    getSession()
    

  }
},[])

//Session Control
useEffect(() => {
  if (!session) {
    setSession(uuidv4());
    sessionStorage.setItem('session', JSON.stringify(uuidv4()));
  }
}, [session]);

//Show message when the user make a registration
useEffect(() => {
  
  if(isRegistered){
    setTimeout(() => {
      setIsRegistered(false);
    }, 3000)
  }


},[isRegistered])

const handleVerification = () => {
  const getName = document.getElementById("name").value
  const getSector = Array.from(document.getElementById("sector").selectedOptions).map(({ value }) => value);
  const getTerms = document.getElementById("terms").checked
  const newStatus = {...verification }
  newStatus.name = getName.length > 0 ? true : false
  newStatus.sector = getSector.length > 0 ? true : false
  newStatus.terms = getTerms
  const {name,sector,terms} = newStatus
  newStatus.error = (name === false || sector === false || terms === false) ? true : false;
  if(getTerms) {setIsDisabled(false)}
  setVerification(newStatus)

}

const handleClick = () => {
  const name = document.getElementById("name").value
  const sector = Array.from(document.getElementById("sector").selectedOptions).map(({ value }) => value);

  handleVerification()
  

  if(!error){
    const data = {
      "userId": name,
      "fieldId": [sector],
      "session": session
    }
    sendToRegister(data)
  }
  
}

const handleValidCharacters = () => {
  const str = document.getElementById("name").value;
  const hasSpecialChar = Regex(str, "test.NoEspecialCharacters");
  const newStatus = {...verification}
  newStatus.char = hasSpecialChar
  setVerification(newStatus); 
}

const {error, name, sector, terms} = verification;

  return (
  <>
  <div className="home">
    <div className="container">
      <label className="title"> Please enter your name and pick the sectors you are currently involved in </label>
      {error && name === false && <label className="error" data-testid="nameError" >Required field</label>} 
      {verification.char && <label className="error" data-testid="nameErrorChar">Special characters are not allowed.</label>} 
      <div className={`name`}>
        <input type="text" name="name" data-testid="name" className={`${verification.char && 'errorChar'}`} onChange={() =>  handleValidCharacters() } id="name"/> 
      </div>
      {error && sector === false &&<label className="error" data-testid="sectorError">Required field</label>} 
      <div className="sector" > 
        <select size="5" name="sector" data-testid="sector" id="sector" multiple> 
          { sectors && sectors.map(sector => <option value={sector.id}  key={sector.id}>{sector.field}</option> )}
        </select> 
        {isRegistered && <div className="isRegistered"> The data was successfully saved </div>} 
      </div>

      {error && terms === false &&<label className="error" data-testid="termsError" >Required field</label>} 
      <div className="terms">
        <input type="checkbox" data-testid="terms" id="terms" onClick={() => handleVerification() } name="terms" />
        <label>Agree to terms</label>
      </div>  
      <input type="submit" className="submit" data-testid="submit" onClick={() =>  handleClick() } value="Save" disabled={isDisabled}   />
    </div>
  </div>
  </>
)}