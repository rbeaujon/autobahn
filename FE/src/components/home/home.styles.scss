$md-bg:  #2196F3;
$md-sectors-bg: #FAFAFA; 
$md-sectors-options: #212121; 
$md-button: #FFB300;
$md-text: #212121;

$mq-width-80: 80%;
$mq-width-40: 40%;

.home {
  position: relative;
  color: $md-text;
  .container {
    position: absolute;
    background-color: $md-bg; 
    transform: translate(-50%, 5%);
    top: 10%;
    left: 50%;
    width: 80%;
    height: 80vh;
    padding: 30px;
    display: flex;
    flex-direction: column;
    align-items: center;
    box-shadow:  0 10px 30px 10px rgba(0, 0, 0, 0.268);

    .title {
      color: #424242;
      font-size: 1.5em;
      font-weight: bold;
      margin-bottom: 20px;
      width: 40%;
      text-align: justify;
    }

    .name {
      margin-bottom: 20px;
      width: 100%; 
      text-align: center;
      input {
        height: 20px;
      }    
    }

    .sector {
      margin-bottom: 20px;
      select {
        width: 100%;
        height: 200px;
        background-color: $md-sectors-bg;
      }
      option {
        color: $md-sectors-options;
      }
      label {
        color: $md-text;
        font-size: 0.8em;
      } 
    }

  
    .terms {
      margin-bottom: 20px;
    }

    .submit {
      background-color: $md-button;
      border: none;
      padding: 10px 20px;
      border-radius: 5px;
      cursor: pointer;
      font-size: 1em;
    }

    .error {
      color: #F44336;
      font-weight: 800;
      font-size: 0.8em;
      margin-bottom: 5px;
      text-transform:lowercase ;
    }
    .errorChar{
      border: 2px solid #F44336;
    }
    .isRegistered{
      background-color:  #7CB342;
      color: white;
      font-size: 0.7em;
      font-weight: 600;
      text-transform:uppercase ;
      margin-top: 10px;
      padding: 5px;
      text-align: center;
    }

  }
}

/* Mobiles */
@media only screen and (max-width: 320px) {
  .home .container {
    width: $mq-width-80;
    height: auto;
    padding: 15px;

    .title, .name input, .sector, .submit, .error, .errorChar {
      width: $mq-width-80;
    }
    .title {
      font-size: 0.9em;
    }
    .sector {
      select {
        height: 25vh;
      }
    }
    .submit {
      font-size: 0.9em;
      padding: 5px 10px;
    }
    .error {
      font-size: 0.7em;
    }
  }
}

@media only screen and (min-width: 321px) and (max-width: 767px) {
  .home .container {
    width: $mq-width-80;
    height: auto;
    padding: 15px;

    .title, .name input, .sector, .submit, .error, .errorChar {
      width: $mq-width-80;
    }
    .title {
      font-size: 1em;
    }
    .sector {
      select {
        height: 35vh;
      }
    }
    .submit {
      font-size: 0.9em;
    }
    .error {
      font-size: 0.7em;
    }
  }
}

/* Tablet */
@media only screen and (min-width: 768px) and (max-width: 991px) {
  .home .container {
    width: 60%;

    .title, .name input, .sector, .submit, .error, .errorChar {
      width: $mq-width-80;
    }
    .title {
      font-size: 2em;
    }
    .sector {
      select {
        height: 40vh;
      }
    }
    .submit {
      font-size: 1.2em;
    }
    .error {
      font-size: 1em;
    }
  }
}

/* Laptop */
@media only screen and (min-width: 992px) {
  .home .container {
    width: $mq-width-40;

    .title, .name input, .sector, .submit, .error, .errorChar {
      width: $mq-width-80;
    }

    .sector {
      select {
        height: 40vh;
      }
    }
  }
}
