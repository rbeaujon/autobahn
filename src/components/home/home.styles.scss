$md-bg:  #2196F3;
$md-sectors-bg: #FAFAFA; 
$md-sectors-options: #212121; 
$md-button: #FFB300;
$md-text: #212121;

.home {
  position: relative;
  color: $md-text;
  .container {
    position: absolute;
    background-color: $md-bg; 
    transform: translate(-50%, 5%);
    top: 10%;
    left: 50%;
    width: 80%;
    height: 80vh;
    padding: 30px;
    display: flex;
    flex-direction: column;
    align-items: center;
    box-shadow:  0 10px 30px 10px rgba(0, 0, 0, 0.268);

    .title {
      color: #424242;
      font-size: 1.5em;
      font-weight: bold;
      margin-bottom: 20px;
      width: 40%;
    }

    .name {
      margin-bottom: 20px;
      width: 100%; 
      text-align: center;
      input {
        width: 40%;
        height: 20px;
      }
      
    }

    .sector {
      margin-bottom: 20px;
      width: 40%;
      select {
        width: 100%;
        height: 200px;
        background-color: $md-sectors-bg;
      }
      option {
        color: $md-sectors-options;
      }
      label {
        color: $md-text;
        font-size: 0.8em;
      } 
    }

  
    .terms {
      margin-bottom: 20px;
    }

    .submit {
      background-color: $md-button;
      border: none;
      padding: 10px 20px;
      border-radius: 5px;
      cursor: pointer;
      width: 40%;
      font-size: 1em;
    }

    .error {
      color: #F44336;
      font-weight: 800;
      font-size: 0.8em;
      margin-bottom: 5px;
      text-transform:lowercase ;
      width: 40%;
    }
    .errorChar{
      border: 2px solid #F44336;
      width: 40%;
    }
    .isRegistered{
      background-color:  #7CB342;
      color: white;
      font-size: 0.7em;
      font-weight: 600;
      text-transform:uppercase ;
      margin-top: 10px;
      padding: 5px;
      text-align: center;
    }
  }
}

