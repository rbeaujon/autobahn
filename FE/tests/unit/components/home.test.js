import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {render, fireEvent, screen} from '@testing-library/react';
import {Home} from '../../../src/components/home/home';


describe('Home', () => {

  it('Should match snapshot', () => {
    const component = <Home />; 
    expect(component).toMatchSnapshot();
  });
  
  it('Should have the correct initial state values', () => {
    
    render(<Home />);
  
    const sectorsMock = [
      {id: 1, fieldId: 'Services'},
      {id: 2, fieldId: 'Other'},
      {id: 3, fieldId: 'Control'},
      {id: 4, fieldId: 'Domain'},
      {id: 5, fieldId: 'Engineering'},
      {id: 6, fieldId: 'Frontend'},
      {id: 7, fieldId: 'Groups'},
      ];
    
    const sector = screen.getByTestId("sector");
    
    sectorsMock.map(elem => {
      let option = document.createElement("option")
      option.text =elem.fieldId
      sector.add(option)
    });
    
    expect(sector.options.length).toBe(7);
    expect(sector.options[0].textContent).toBe("Services");
  });

  it('Should the title be renders correctly', () => {
    render(<Home />);
    const titleElement = screen.getByText('Please enter your name and pick the sectors you are currently involved in');
    expect(titleElement).toBeTruthy();
  });

  it("should call the API and set the sectors state", async () => {
    
    const expectedData = [{id:1, field:"Service"},{id:2, field:"Printing"}];

    const mockSectorsAPI = jest.fn();
    jest.mock('../../../src/services/API/sectors.api', () => ({
      SectorsAPI: mockSectorsAPI,
    }));
    mockSectorsAPI.mockResolvedValue({
      json: expectedData,
      status: 200, 
      ok: true
    });

    const data = await mockSectorsAPI();

    expect(mockSectorsAPI).toHaveBeenCalledTimes(1);
    expect(data.json[0].field).toEqual('Service');
});

  it('Should call handleValidCharacters when input changes', () => {
    
    render(<Home/>);

    const input = screen.getByTestId('name');
    const terms = screen.getByTestId('terms');

    fireEvent.change(input, { target: { value: 'John Doe%^&$', text:"@#!@#!" } });
    terms.checked = true
    const error = screen.getByTestId('nameErrorChar');
    
    expect(error.textContent).toEqual('Special characters are not allowed.');
  });
  
  it('should call handleClick when submit button is clicked', () => {
    const handleClick = jest.fn();
    render(<Home handleClick={handleClick}/>);

    const submitButton = screen.getByTestId('submit');
   
    submitButton.onclick=handleClick
    fireEvent.click(submitButton);
   
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('Should show required when is clicked on the submit button without data', () => {

    render(<Home/>);

    const terms = screen.getByTestId('terms');
    const submit = screen.getByTestId('submit');

    fireEvent.click(terms)
    fireEvent.click(terms)
    fireEvent.click(submit)

    const nameError = screen.getByTestId('nameError');    
    const sectorError = screen.getByTestId('sectorError');
    const termsError = screen.getByTestId('termsError');

    expect(nameError.textContent).toEqual('Required field');
    expect(sectorError.textContent).toEqual('Required field');
    expect(termsError.textContent).toEqual('Required field');
  })

  it('handleClick function works correctly', () => {
    
    render(<Home />);

    // Trigger the handleClick function
    const nameInput = screen.getByTestId('name');
    nameInput.value = 'Test user';
    const sectorInput = screen.getByTestId('sector');
    sectorInput.value = 'Test sector';
    const submit = screen.getByTestId('submit');
    submit.click();

    // Verify the function works correctly
    // Add your own assertions here to check the values of the name and sector variables
  });

});
