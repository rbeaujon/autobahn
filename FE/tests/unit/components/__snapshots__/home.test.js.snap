// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should match snapshot 1`] = `<Home />`;

